src/tools.ts

```diff
@@ -90,41 +67,103 @@ export function addMathLink(outLinkEl: HTMLElement, mathLink: string, newElement
 
     finishRenderMath();
     if (newElement) {
-        outLinkEl.parentNode.insertBefore(mathLinkEl, outLinkEl.nextSibling);
+        targetEl.parentNode.insertBefore(mathLinkEl, targetEl.nextSibling);
         mathLinkEl.classList.add("mathLink-internal-link");
-        outLinkEl.classList.add("original-internal-link");
-        outLinkEl.style.display = "none";
+        targetEl.classList.add("original-internal-link");
+        targetEl.style.display = "none";
     }
 
     return mathLinkEl;
 }
 
-export function getMathLink(plugin: MathLinks, file: TFile): string {
-    if (!file) return undefined;
-
-    let mathLink = plugin.app.metadataCache.getFileCache(file)?.frontmatter?.mathLink;
+// Get mathLink from cached metadata
+export function getMathLink(plugin: MathLinks, targetLink: string, sourcePath: string): string {
+    let { path, subpath } = parseLinktext(targetLink);
+
+    let file = plugin.app.metadataCache.getFirstLinkpathDest(path, sourcePath);
+    if (!file) return "";
+
+    let cache = plugin.app.metadataCache.getFileCache(file);
+    if (!cache) return "";
+
+    let mathLink = "";
+    if (cache.frontmatter) {
+        let subpathResult = resolveSubpath(cache, subpath);
+        if (subpathResult) {
+            let subMathLink = "";
+            if (subpathResult.type == "heading") {
+                subMathLink = subpathResult.current.heading;
+            } else if (subpathResult.type == "block" && cache.frontmatter["mathLinks-block"]) {
+                subMathLink = "^" + cache.frontmatter["mathLinks-block"][subpathResult.block.id];
+            }
 
-    if (mathLink == "auto") {
-        let templates = plugin.settings.templates;
-        mathLink = file.name.replace("\.md", "");
-        for (let i = 0; i < templates.length; i++) {
-            let replaced = new RegExp(templates[i].replaced);
-            let replacement = templates[i].replacement;
+            if (path) { 
+                mathLink = (cache.frontmatter.mathLink ?? path) + " > " + subMathLink;
```

Comment:
This was `... + " # " + ...`, but I've tested it when MathLinks disabled and it renders as ` ... > ...` in reading mode. I've changed it accordingly (and in live-preview too)

Reply:
> This was `... + " # " + ...`, but I've tested it when MathLinks disabled and it renders as ` ... > ...` in reading mode. I've changed it accordingly (and in live-preview too)

Thanks, I misunderstood. 

By the way, it might be helpful to let users choose how the heading/block links are rendered. 
For example, maybe some people would like `Definition (foo)` rather than `^Definition (foo)` or `title > ^Definition (foo)`. 
But I'm not sure how this should be realized. It's just a rough thought.

Reply:
> By the way, it might be helpful to let users choose how the heading/block links are rendered.
> For example, maybe some people would like `Definition (foo)` rather than `^Definition (foo)` or `title > ^Definition (foo)`.
> But I'm not sure how this should be realized. It's just a rough thought.

This is a good idea! I'll leave it for a later update though, since this one will be quite packed already, and adding this open requires some work in `settings.ts`, which I have not touched in a long time! I'll mark it in a new discussion.